cpp/src/arrow/engine/substrait/options.cc

```diff
@@ -0,0 +1,118 @@
+                                  const google::protobuf::Any& rel,
+                                  const ExtensionSet& ext_set) override {
+    if (rel.Is<arrow::substrait::AsOfJoinRel>()) {
+      arrow::substrait::AsOfJoinRel as_of_join_rel;
+      rel.UnpackTo(&as_of_join_rel);
+      return MakeAsOfJoinRel(inputs, as_of_join_rel, ext_set);
+    }
+    return Status::NotImplemented("Unrecognized extension in Susbstrait plan: ",
+                                  rel.DebugString());
+  }
+
+ private:
+  Result<DeclarationInfo> MakeAsOfJoinRel(
+      const std::vector<DeclarationInfo>& inputs,
+      const arrow::substrait::AsOfJoinRel& as_of_join_rel, const ExtensionSet& ext_set) {
+    if (inputs.size() < 2) {
+      return Status::Invalid("substrait::AsOfJoinNode too few input tables: ",
+                             inputs.size());
+    }
+    if (static_cast<size_t>(as_of_join_rel.keys_size()) != inputs.size()) {
+      return Status::Invalid("substrait::AsOfJoinNode mismatched number of inputs");
+    }
+
+    size_t n_input = inputs.size(), i = 0;
+    std::vector<compute::AsofJoinNodeOptions::Keys> input_keys(n_input);
+    for (const auto& keys : as_of_join_rel.keys()) {
+      // on-key
+      if (!keys.has_on()) {
+        return Status::Invalid("substrait::AsOfJoinNode missing on-key for input ", i);
+      }
+      ARROW_ASSIGN_OR_RAISE(auto on_key_expr, FromProto(keys.on(), ext_set, {}));
+      if (on_key_expr.field_ref() == NULLPTR) {
+        return Status::NotImplemented(
+            "substrait::AsOfJoinNode non-field-ref on-key for input ", i);
+      }
+      const FieldRef& on_key = *on_key_expr.field_ref();
+
+      // by-key
+      std::vector<FieldRef> by_key;
+      for (const auto& by_item : keys.by()) {
+        ARROW_ASSIGN_OR_RAISE(auto by_key_expr, FromProto(by_item, ext_set, {}));
+        if (by_key_expr.field_ref() == NULLPTR) {
+          return Status::NotImplemented(
+              "substrait::AsOfJoinNode non-field-ref by-key for input ", i);
+        }
+        by_key.push_back(*by_key_expr.field_ref());
+      }
+
+      input_keys[i] = {std::move(on_key), std::move(by_key)};
+      ++i;
+    }
+
+    // schema
+    int64_t tolerance = as_of_join_rel.tolerance();
+    std::vector<std::shared_ptr<Schema>> input_schema(inputs.size());
+    for (size_t i = 0; i < inputs.size(); i++) {
+      input_schema[i] = inputs[i].output_schema;
+    }
+    ARROW_ASSIGN_OR_RAISE(auto schema,
+                          compute::asofjoin::MakeOutputSchema(input_schema, input_keys));
+    compute::AsofJoinNodeOptions asofjoin_node_opts{std::move(input_keys), tolerance};
+
+    // declaration
+    std::vector<compute::Declaration::Input> input_decls(inputs.size());
+    for (size_t i = 0; i < inputs.size(); i++) {
+      input_decls[i] = inputs[i].declaration;
+    }
+    return DeclarationInfo{
+        compute::Declaration("asofjoin", input_decls, std::move(asofjoin_node_opts)),
+        std::move(schema)};
+  }
```

Comment:
There is a proposal on https://github.com/ibis-project/ibis-substrait/issues/423 to fix up the output ordering to something ibis-substrait expects.  I think we can handle it as part of this PR or as a separate follow-up PR.  Let me know which you'd prefer.

Reply:
I'd prefer a separate follow-up PR.